### DISCUSSION TITLE
langchain search returns "out of range in json: ef(10) should be larger than k(20)" **only after running for some time/overnight**

### DISCUSSION THREAD

**Discussion by dammad** [milvus-io/milvus#35160] [Q&A and General discussion]
We've been using Langchain and Milvus for a chatbot project and it has been running fine. Recently we noticed that if Milvus is left overnight with a 2009 item collection, the queries fail with "failed to search: out of range in json: ef(10) should be larger than k(20))". 

I've been accessing Milvus with:
 ```
  # Setup the langchain access to Milvus as a retriever
    milvusretriever = vector_store.as_retriever(search_kwargs={"k": 20, "ef":30})
```

So EF is large than top-K. Anyone any ideas what would cause Milvus to be working fine for hours and then do this? The only way to resolve is to re-create the collection, rebuilding the index doesn't help. 

Error message:

`"MilvusException()Traceback (most recent call last):\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/runnables/base.py\", line 2507, in invoke\n    input = step.invoke(input, config)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/retrievers.py\", line 221, in invoke\n    raise e\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/retrievers.py\", line 214, in invoke\n    result = self._get_relevant_documents(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain/retrievers/multi_query.py\", line 169, in _get_relevant_documents\n    documents = self.retrieve_documents(queries, run_manager)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain/retrievers/multi_query.py\", line 207, in retrieve_documents\n    docs = self.retriever.invoke(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/retrievers.py\", line 221, in invoke\n    raise e\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/retrievers.py\", line 214, in invoke\n    result = self._get_relevant_documents(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain/retrievers/contextual_compression.py\", line 44, in _get_relevant_documents\n    docs = self.base_retriever.invoke(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/retrievers.py\", line 221, in invoke\n    raise e\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/retrievers.py\", line 214, in invoke\n    result = self._get_relevant_documents(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_core/vectorstores.py\", line 797, in _get_relevant_documents\n    docs = self.vectorstore.similarity_search(query, **self.search_kwargs)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_community/vectorstores/milvus.py\", line 663, in similarity_search\n    res = self.similarity_search_with_score(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_community/vectorstores/milvus.py\", line 736, in similarity_search_with_score\n    res = self.similarity_search_with_score_by_vector(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/langchain_community/vectorstores/milvus.py\", line 781, in similarity_search_with_score_by_vector\n    res = self.col.search(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/orm/collection.py\", line 799, in search\n    resp = conn.search(\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/decorators.py\", line 147, in handler\n    raise e from e\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/decorators.py\", line 143, in handler\n    return func(*args, **kwargs)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/decorators.py\", line 182, in handler\n    return func(self, *args, **kwargs)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/decorators.py\", line 122, in handler\n    raise e from e\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/decorators.py\", line 87, in handler\n    return func(*args, **kwargs)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/client/grpc_handler.py\", line 801, in search\n    return self._execute_search(request, timeout, round_decimal=round_decimal, **kwargs)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/client/grpc_handler.py\", line 742, in _execute_search\n    raise e from e\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/client/grpc_handler.py\", line 735, in _execute_search\n    check_status(response.status)\n\n\n  File \"/home/ubuntu/.local/lib/python3.10/site-packages/pymilvus/client/utils.py\", line 63, in check_status\n    raise MilvusException(status.code, status.reason, status.error_code)\n\n\npymilvus.exceptions.MilvusException: <MilvusException: (code=65535, message=fail to search on QueryNode 17: worker(17) query failed:  => failed to search: out of range in json: ef(10) should be larger than k(20))>`

**Comment by yhmo** [comment]
In langchain source code:
- as_retriever() declares a VectorStoreRetriever object with **search_kwargs**:
https://github.com/langchain-ai/langchain/blob/f14121faafe0366e8b472b71159fea138308d4ed/libs/core/langchain_core/vectorstores/base.py#L1202

- VectorStoreRetriever._get_relevant_documents() passes the **search_kwargs** to vectorstore.similarity_search()
https://github.com/langchain-ai/langchain/blob/f14121faafe0366e8b472b71159fea138308d4ed/libs/core/langchain_core/vectorstores/base.py#L1248

- Milvus VectorStore disassembles the **search_kwargs** to be parameters including: query, k, param, expr, etc.
https://github.com/langchain-ai/langchain/blob/f14121faafe0366e8b472b71159fea138308d4ed/libs/community/langchain_community/vectorstores/milvus.py#L635

- similarity_search_with_score_by_vector() passes the query, k, param, expr to milvus interface collection.search()
https://github.com/langchain-ai/langchain/blob/f14121faafe0366e8b472b71159fea138308d4ed/libs/community/langchain_community/vectorstores/milvus.py#L781

param is a dict that contains the index search parameters such as nprobe, ef.
So, the correct format of **search_kwargs** should be: search_kwargs={"k": 20, "param": {"ef":30}}

**Comment by dammad** [comment]
Got it! But does that explain it works for many hours and then stopping working?

**Reply by yhmo** [reply]
The ef/k checker begins from a minor version of 2.2.x or 2.3.x. In the legacy version, k can be larger than ef. I guess your milvus version is upgraded.

**Reply by dammad** [reply]
I thought that too, but when I setup a fresh collection on the same Milvus is works fine for hours.. Then, typically overnight, something happens that causes it to throw up these errors.

**Reply by xiaofan-luan** [reply]
what is your milvus version?
If we have logs that could be really helpful

make sure you are using 2.3.20 or 2.4.7
